docs/decisions/0011-producing-to-event-bus-via-settings.rst

```diff
@@ -0,0 +1,106 @@
+
+Create a generic signal handler to push events to the event bus. This handler
+should be attached to or connected to the signals that are enabled in Django settings.
+of the application. The configuration format will be as shown below:
+
+.. code-block:: python
+
+   OPEN_EDX_EVENTS_CONFIG = {
+       'org.openedx.content_authoring.xblock.deleted.v1': {
+           'topics': 'xblock-status',
+           'event_key_field': 'xblock_info.usage_key',
+           'event_data_key': 'xblock_info',
+           'enabled': True,
+       },
+       'org.openedx.content_authoring.xblock.published.v1': {
+           'topics': ['xblock-status', 'xblock-published'],
+           'event_key_field': 'xblock_info.usage_key',
+           'event_data_key': 'xblock_info',
+           'enabled': True,
+       }
+   }
+
+This configuration will be read in openedx_events ``apps.OpenedxEventsConfig(AppConfig).ready``
+method and a generic signal handler will be connected to the event_types (keys)
+listed in the configuration after validating its format.
+
+.. code-block:: python
+
+   def ready(self):
+       load_all_signals()
+       signals_config = getattr(settings, "OPEN_EDX_EVENTS_CONFIG", {})
+       # validate signals_config before moving to next step
+       for event_type, config in signals_config.items():
+           signal = OpenEdxPublicSignal.get_signal_by_type(event_type)
+           if config.get("enabled", True):
+               signal.connect(general_signal_handler, dispatch_uid=event_type)
+           else:
+               signal.disconnect(general_signal_handler, dispatch_uid=event_type)
+       return super().ready()
+
+The generic handler will again read the configuration and get details for
+event or signal triggered and push it to the event bus.
+
+.. code-block:: python
+
+   def general_signal_handler(sender, signal, **kwargs):
+       signal_config = getattr(settings, "OPEN_EDX_EVENTS_CONFIG", {}).get(signal.event_type)
+       if signal_config:
+           if isinstance(signal_config["topics"], str):
+               topics = [signal_config['topics']]
+           else:
+               topics = signal_config['topics']
+           for topic in topics:
+               get_producer().send(
+                   signal=signal,
+                   topic=topic,
+                   event_key_field=signal_config['event_key_field'],
+                   event_data={signal_config['event_data_key']: kwargs[signal_config['event_data_key']]},
+                   event_metadata=kwargs['metadata'],
+               )
+
+Consequences
+************
+* Applications producing to event_bus can push events without requiring any
+  additional code.
+* The configuration is a dictionary, which makes it flexible but difficult to
+  enforce schemas and maintain them.
+* Producing as well as consuming applications will be required to add
+  ``openedx_events`` as an application in their ``INSTALLED_APPS`` settings.
+
+Rejected Alternatives
+*********************
+
+None
```

Comment:
It's probably worth a sentence or two describing the downsides of the current approach, which is an alternative that we're rejecting.

Reply:
Done.